crates/sui-network/src/state_sync/mod.rs

```diff
@@ -556,33 +552,81 @@ async fn notify_peers_of_checkpoint(
     futures::future::join_all(futs).await;
 }
 
-async fn get_latest_from_peer(peer: anemo::Peer, peer_heights: Arc<RwLock<PeerHeights>>) {
+async fn get_latest_from_peer(
+    our_genesis_checkpoint_digest: CheckpointDigest,
+    peer: anemo::Peer,
+    peer_heights: Arc<RwLock<PeerHeights>>,
+) {
     let peer_id = peer.peer_id();
-    let request = Request::new(GetCheckpointSummaryRequest::Latest).with_timeout(DEFAULT_TIMEOUT);
-    let response = StateSyncClient::new(peer)
-        .get_checkpoint_summary(request)
-        .await
-        .map(Response::into_inner);
-    update_peer_height(&peer_heights, peer_id, &response);
-}
+    let mut client = StateSyncClient::new(peer);
 
-fn update_peer_height(
-    peer_heights: &RwLock<PeerHeights>,
-    peer_id: PeerId,
-    response: &Result<Option<Checkpoint>, Status>,
-) {
-    match response {
-        Ok(latest) => {
-            peer_heights
-                .write()
-                .unwrap()
-                .update_peer_height(peer_id, latest.clone());
-        }
-        Err(status) => {
-            trace!("get_latest_checkpoint_summary request failed: {status:?}");
-            peer_heights.write().unwrap().heights.remove(&peer_id);
+    let info = {
+        let maybe_info = peer_heights.read().unwrap().peers.get(&peer_id).copied();
+
+        if let Some(info) = maybe_info {
+            info
+        } else {
+            // TODO do we want to create a new API just for querying a node's chainid?
+            //
+            // We need to query this node's genesis checkpoint to see if they're on the same chain
+            // as us
+            let request = Request::new(GetCheckpointSummaryRequest::BySequenceNumber(0))
+                .with_timeout(DEFAULT_TIMEOUT);
+            let response = client
+                .get_checkpoint_summary(request)
+                .await
+                .map(Response::into_inner);
+
+            let info = match response {
+                Ok(Some(checkpoint)) => {
+                    let digest = checkpoint.digest();
+                    PeerStateSyncInfo {
+                        genesis_checkpoint_digest: digest,
+                        on_same_chain_as_us: our_genesis_checkpoint_digest == digest,
+                        height: checkpoint.sequence_number(),
+                    }
+                }
+                Ok(None) => PeerStateSyncInfo {
+                    genesis_checkpoint_digest: CheckpointDigest::default(),
+                    on_same_chain_as_us: false,
+                    height: CheckpointSequenceNumber::default(),
+                },
+                Err(status) => {
+                    trace!("get_latest_checkpoint_summary request failed: {status:?}");
+                    return;
+                }
+            };
+            peer_heights.write().unwrap().peers.insert(peer_id, info);
```

Comment:
I think a re-check might also be needed here to verify that you aren't overwriting with an old info? Since you unlocked already after reading `maybe_info` above. unless there's something else that guarantees no overlapping calls to `get_latest_from_peer` for the same peer? (if so maybe document that with a comment on the map write, for posterity?)

Reply:
I've added a method that will do this extra checking on insert and makes a decision on what to do if there was a race.